### PR TITLE
Bug/fix eslint

### PR DESCRIPTION
**Pull request by JGets**
- fixes the `eslintrc` configuration
- fixes all warnings & errors generated by eslint

### CODE REVIEW

**Comment by JGets** [comment]
[JS.js has a lot of eslint errors](https://trello.com/c/y6YiLQfn/8590-jsjs-has-a-lot-of-eslint-errors)

File: JS.js [hunk 1]
```diff
@@ -1,1162 +1,1260 @@
...
 					if (!this.hasOwnProperty('__abstract')) {
-						this.__abstract           = false;
-						this.__abstractMethodName = undefined;
-						this.__abstractFieldName  = undefined;
+						const implementedMethods = {};
+						const implementedFields  = {};
+						let currentClass       = this;
+
+						while (currentClass && currentClass.properties && !this.hasOwnProperty('__abstract')) {
+							// COULDDO: don't repeat these blocks - makes it a little more difficult to differentiate
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: !==

Comment (JoelHobsonRoadmunk) [inline JS.js]: Align

File: test/mixin.js [hunk 2]
```diff
@@ -129,164 +135,182 @@ describe('mixin', function() {
...
+					return 'method4';
+				},
+			},
 		});
 
 		expect(Base.isAbstract()).to.be.true;
-		expect(Sub .isAbstract()).to.be.false;
+		expect(Sub.isAbstract()).to.be.false;
 
-		var sub  = new Sub();
```
Comment (JoelHobsonRoadmunk) [inline test/mixin.js]: Extra space before =

File: JS.js [hunk 3]
```diff
@@ -1,1162 +1,1260 @@
...
-		// check for stub call
-		if (originalProperties === undefined)
-			return clazz;
-	}
-	else if (typeof className === "function") {
-		clazz     = className;
-		className = clazz.__className__;
-	}
-	else throw new Error("invalid parameter className: " + className);
+		if (typeof className == 'string') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 4]
```diff
@@ -1,1162 +1,1260 @@
...
-	var processedFields = {}, unprocessedFields, fieldName;
-	var properties = clazz.properties;
+		return clazz;
+	};
 
-	do {
-		unprocessedFields = {};
+	function setupSuperClass(clazz, properties) {
+		// setup super class
+		if (properties.inherits && typeof properties.inherits != 'function') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: !==

File: JS.js [hunk 5]
```diff
@@ -1,1162 +1,1260 @@
...
+		// done in a loop cause the event hander could add new field definitions
+		let unprocessedFields, fieldName;
+		const processedFields = {};
+		const properties      = clazz.properties;
+
+		do {
+			unprocessedFields = {};
+
+			for (fieldName in properties.fields) {
+				if (properties.fields[fieldName] != processedFields[fieldName]) {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: !==

File: JS.js [hunk 6]
```diff
@@ -1,1162 +1,1260 @@
...
-			else if (typeof field.type != "function") {
-				var a = normalizeField(field.type);
-				field.type = a.type;
-				field.init = field.init === undefined ? a.init : field.init;
+
+			for (fieldName in unprocessedFields) {
+				invokeEvent('fieldDefinition', [ clazz, fieldName, unprocessedFields[fieldName] ]);
+
+				// check if entire field has been replaced
+				if (properties.fields[fieldName] != unprocessedFields[fieldName]) {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: !==

File: JS.js [hunk 7]
```diff
@@ -1,1162 +1,1260 @@
...
+					if (!field.hasOwnProperty('init')) {
+						field.init = '';
+					}
+				}
+				else if (field.type === Number) {
+					if (!field.hasOwnProperty('init')) {
+						field.init = 0;
+					}
+				}
+				else if (typeof field.type != 'function') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: !==

File: JS.js [hunk 8]
```diff
@@ -1,1162 +1,1260 @@
...
 
-	// 1. add from any base classes
-	extendPropertiesWithClass(properties, classProperties.inherits, {
-		isMixin     : options.isMixin,
-		copyMethods : options.isMixin && options.copyMethods
-	});
+		const classProperties = originalPropertiesMap.get(clazz);
+		if (!classProperties) {
+			// check if clazz is not a JS.class class
+			if (clazz == BaseClass || BaseClass.isPrototypeOf(clazz)) {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 9]
```diff
@@ -1,1162 +1,1260 @@
...
 
-			else {
-				var method = makeWrapper(sourceMethod);
-				method.methodType = "method";
-				method.methodName = methodName;
-				method.__class__  = clazz;
-				dest[methodName] = method;
+				else {
+					const method = makeWrapper(sourceMethod);
+					method.methodType = 'method';
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: Align

File: JS.js [hunk 10]
```diff
@@ -1,1162 +1,1260 @@
...
+	 * @param {object} options is an optional list of additional parameters:
+	 *                 {boolean} asFunction, invokes the specified property as a function with the value as the only parameter
+	 */
+	JS.setter = function(property, options) {
+		options = options || {};
+
+		if (options.asFunction) {
+			return function(value) {
+				const prop = this[property];
+				if (typeof prop == 'function') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 11]
```diff
@@ -1,1162 +1,1260 @@
...
+	 * Helper function that creates methods on the class.
+	 * Expects to be called in the context of the class object.
+	 * @param  {Object} definitions the class property definitions object
+	 * @param  {Object} destination where to add the new method
+	 */
+	function createMethods(definitions, destination) {
+		for (const name in definitions) {
+			let method = definitions[name];
+
+			if (typeof method == 'object') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 12]
```diff
@@ -1,1162 +1,1260 @@
...
+					Object.defineProperty(destination, name, descriptor);
 				}
-				if (typeof method.set === "function") {
-					method.set = makeSuperStackUpdaterProxy(method.set);
-					method.set.methodType = "set";
-					method.set.methodName = name;
-					method.set.__class__  = this;
-					method.set.super      = makeSuperMethodProxy(method.set);
-					descriptor.set        = method.set;
+				else if (typeof method.$super == 'function') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 13]
```diff
@@ -1,1162 +1,1260 @@
...
+					method = method.$super.call(this, superMethod);
 				}
+				else if (name[0] === '$') {
+					// it's a region
+					createMethods.call(this, method, destination);
+				}
+			}
 
-				Object.defineProperty(destination, name, descriptor);
+			if (typeof method == 'function') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 14]
```diff
@@ -1,1162 +1,1260 @@
...
-	if (typeof field.get == "function" && typeof field.set == "undefined")
-		return undefined;
+	/**
+	 * Returns the initial value for a field.
+	 * @param  {Object | Function} instance is the instance of the class or the class itself if the field is static
+	 * @param  {Object | String} field the field specification or field name
+	 * @return {*}
+	 */
+	JS.class.initialFieldValue = function(instance, field) {
+		if (typeof field == 'string') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 15]
```diff
@@ -1,1162 +1,1260 @@
...
-			return field.init.call(instance);
-		case "string":
-		case "number":
-		case "boolean":
-			return field.init;
-		case "object":
-			if (field.init === null)
-				return null;
+		// check if this is a getter field (with no setter thus not writable) then do nothing
+		if (typeof field.get == 'function' && typeof field.set == 'undefined') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 16]
```diff
@@ -1,1162 +1,1260 @@
...
+			if (desc) {
+				if (method.methodType == 'method' && typeof desc.value == 'function') {
+					return desc.value;
 				}
-				else {
-					currentStack.pop();
+				if (method.methodType == 'get' && typeof desc.get === 'function') {
+					return desc.get;
 				}
+				if (method.methodType == 'set' && typeof desc.set === 'function') {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 17]
```diff
@@ -1,1162 +1,1260 @@
...
+		return constructionStack.indexOf(instance) >= 0;
+	};
 
-		/**
-		 * Gets called after the instance has been constructed.
-		 * Can be overriden in instance classes to provide post-construction functionality.
-		 */
-		afterCreateInstance : function() {}
-	},
+	Object.defineProperty(JS.class.isUnderConstruction, 'stackSize', { get : function() {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: Should there be some more newlines here? This looks kind of weird

File: JS.js [hunk 18]
```diff
@@ -1,1162 +1,1260 @@
...
-							if (currentClass.properties.fields[fieldName].abstract && implementedFields[fieldName] === undefined) {
-								this.__abstract = true;
-								this.__abstractFieldName = fieldName;
-								break;
-							}
-							implementedFields[fieldName] = true;
+						if (typeof mixin.hasMixin   == 'function' && mixin.hasMixin(mixinClass))   {
+							return true;
+						}
+						if (typeof mixin.isSubclass == 'function' && mixin.isSubclass(mixinClass)) {
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

File: JS.js [hunk 19]
```diff
@@ -1,1162 +1,1260 @@
...
+				 * abstract methods/fields need to be implemented.
+				 */
+				isAbstract : function() {
 					if (!this.hasOwnProperty('__abstract')) {
-						this.__abstract           = false;
-						this.__abstractMethodName = undefined;
-						this.__abstractFieldName  = undefined;
+						const implementedMethods = {};
+						const implementedFields  = {};
+						let currentClass       = this;
```
Comment (JoelHobsonRoadmunk) [inline JS.js]: Align

File: JS.js (hunk 1 above)
Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

Comment (JoelHobsonRoadmunk) [inline JS.js]: !==

Comment (JoelHobsonRoadmunk) [inline JS.js]: !==

Comment (JoelHobsonRoadmunk) [inline JS.js]: ===

**Pull request merged**